### PR TITLE
Benchmark Visualizer: Conjure Native vs Oxide

### PR DESCRIPTION
**Pull request by PedroGGBM**
The following PR contains an interactive **dashboard** using Dash to display relevant statistics comparing solvers from Conjure Native versus Conjure Oxide:
- Num of Nodes (SavileRow)
- Time Elapsed (milliseconds)

All instructions on how to execute the program `app.py` file and run the shell/Rust script to get solver statistics for _exhaustive_ tests are included in a `README.md` file. Please refer to said file for Python environment setup.

_Extensions_:
- Git submodule for _exhaustive_ Conjure tests; be able to update according to changes to these tests
- Inputting `.essence` files to create new models, and change `.param` file values to get relevant testing statistics for each solver
- [Long term] A recommending system based on the which statistics favor which solver depending on the solver

Please note that, although node count for SavileRow is only relevant when comparing same solver across Oxide vs Native (i.e. Minion), other Conjure Native node counts are still included for when there is full support for other Oxide solvers.

_Examples_:
image

image

(Based on Georgii's Issue #248 on a web UI for Native vs Oxide benchmarks)

### CODE REVIEW

**Comment by niklasdewally** [comment]
@PedroGGBM any change compared to the demo, or is this the same?

**Comment by niklasdewally** [comment]
> It is server-required; Oz mentioned the possibility of hosting it on a Github page.

To clarify, Github pages hosts static files (i.e serverless stuff) only.

**Review by niklasdewally — APPROVED** [review]

**Comment by ozgurakgun** [comment]
I think the dynamic server side is a problem for this actually. 

The main goal is checking the correctness of conjure-oxide (starting from conjure-oxide tests) with respect to conjure. This can be done as part of the existing tester and outputs shown as test results. Making a static dashboard page for easy exploration of the test results would be good as well. As it stands I think this PR tries to do too much and not enough at the same time. We need to iterate on the goals of this work before landing the PR.

**Comment by niklasdewally** [comment]
@ozgurakgun @PedroGGBM 

If static hosting is a requirement, I think Quarto (what used to be Rmarkdown but now for Python too) has static dashboards now that are fairly plotting library agnostic?

https://quarto.org/docs/dashboards/deployment.html#static-dashboards

**Comment by niklasdewally** [comment]
It also has observableJS which might(?) be interactive and serverless, otherwise its generally a good tool for turning code notebooks (Jupyter or R) into reports and websites.

**Comment by niklasdewally** [comment]
https://jjallaire.github.io/gapminder-dashboard/
https://github.com/jjallaire/gapminder-dashboard/blob/main/gapminder.qmd

**Comment by PedroGGBM** [comment]
Thank you @niklasdewally for the Quarto recommendation :)  Just added support for the static HTML dashboard generation, and edited README.md and build.sh correspondingly; all works like a charm.

@ozgurakgun apologies for the delay in getting this in. I suppose now it has a both a dynamic and static dashboard... I'll include this in my final proj report as well.

**Comment by niklasdewally** [comment]
Seems to be a step in the right direction. Here's some of my thoughts about this in general  - these are not necessarily problems with the PR or things that can be done this semester, just things that come to mind: 

* Quarto supports plotly - it is basically a jupyter notebook, so if it works in Jupyter it works here. You could probably put the plotly code inside the notebook itself instead of an image. (I'm sure you know this and did it with images as a first step, just documenting our conversation earlier)
*  Quarto also supports jupyter widgets (as it uses Jupyter as its execution engine) - I am fairly sure Jupyter widgets exist that do the dynamic dropdown with list of tests and so on as Dash does. @ozgurakgun is a notebooks user so might know some useful Jupyter widgets that can go on this? Maybe https://plotly.com/python/figurewidget/ ?
* Pretty sure R plotly can do dropdowns in Javascript, so there is probably a way in Python too (https://plotly.com/python/dropdowns/?).
* I don't want to generate the data all myself on my laptop - is there an example or any images?
*  I would gitignore image files and data - it took a while to clone and the diff crashes my Github.
* This might be useful as well https://quarto.org/docs/computations/parameters.html
* Whats the deployment landscape for this? Also is the data being cached somewhere?
*

**Comment by PedroGGBM** [comment]
Thank you @niklasdewally for the thoughts! The whole Jupyter widgets extension might be a good project to do over the summer or next semester. The images as stored in `./tools/benchmarks-visualizer/figures`. I just realized I didn't add this dir to the .gitignore, but I presume you can see them now. 

As for the deployment landscape, right now it's limited to running the `build.sh`, which allows the user to re-generate the solution stats for each test respective to different solvers. All solution stats are stored in `./data` directory, and all tests (_exhaustive_ tests originally in Conjure repo) are in `./tests`. Git submodules gave loads of problems, so I directly copied all the tests to conjure-oxide (I hope that's ok; they can be used as integration tests in the future). All of this should be detailed in the `README.md` file for easy follow-through.

Thanks again! These are some really good further development ideas.

File: tools/benchmarks-visualizer/tests/exhaustive/README.md
Comment (ozgurakgun) [inline tools/benchmarks-visualizer/tests/exhaustive/README.md:1]: which subset of tests are we importing? not sure about wholesale import of a large number of tests before we are happy with the testing functionality itself. maybe import a few as part of this PR and we can easily add more later once the scaffolding is done?

**Review by ozgurakgun — CHANGES_REQUESTED** [review]

**Comment by ozgurakgun** [comment]
hi @PedroGGBM - if you have time it might be beneficial to talk to @EEJDempster about this work. They are designing the ideal testing plan for this part of the project and your experience would feed into their plan nicely.

**Pull request closed without merging**